File: server/opensandbox_server/api/metrics.py
# Copyright 2026 Alibaba Group Holding Ltd.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""SDK metrics ingestion endpoints."""

from __future__ import annotations

import logging
from typing import Optional

from fastapi import APIRouter, Header, Response, status

from opensandbox_server.api.schema import MetricsEvent
from opensandbox_server.integrations.otel import record_sandbox_create_duration

logger = logging.getLogger(__name__)

router = APIRouter(tags=["Metrics"])


@router.post(
    "/metrics/events",
    status_code=status.HTTP_204_NO_CONTENT,
    responses={
        400: {"description": "Bad request"},
        401: {"description": "Unauthorized"},
        500: {"description": "Internal server error"},
    },
)
def report_metrics_event(
    event: MetricsEvent,
    x_request_id: Optional[str] = Header(None, alias="X-Request-ID"),
) -> Response:
    """Accept best-effort SDK metrics and record to OTEL (noop when disabled)."""
    _ = x_request_id
    if event.event_type == "sandbox.create":
        record_sandbox_create_duration(
            create_duration_ms=event.create_duration_ms,
            sdk_language=event.sdk_language,
            sdk_version=event.sdk_version,
            success=event.success,
        )
        logger.debug(
            "Accepted sandbox.create metrics event sandbox_id=%s image=%s "
            "duration_ms=%s sdk=%s/%s success=%s",
            event.sandbox_id,
            event.image,
            event.create_duration_ms,
            event.sdk_language,
            event.sdk_version,
            event.success,
        )
    return Response(status_code=status.HTTP_204_NO_CONTENT)

File: kubernetes/internal/controller/metrics.go
// Copyright 2026 The OpenSandbox Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package controller

import (
	"context"
	"time"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/metric"
	"go.opentelemetry.io/otel/metric/noop"
	logf "sigs.k8s.io/controller-runtime/pkg/log"

	sandboxv1alpha1 "github.com/alibaba/OpenSandbox/sandbox-k8s/apis/sandbox/v1alpha1"
)

// Allocator-path metrics; see docs/telemetry.md for the signal specification.
const (
	allocatorMeterName = "opensandbox/controller"

	allocatorScheduleDurationMetricName          = "controller.allocator.schedule.duration"
	allocatorPersistAllocStateDurationMetricName = "controller.allocator.persist_alloc_state.duration"
	allocatorSyncAllocResultDurationMetricName   = "controller.allocator.sync_alloc_result.duration"
	allocatorSyncSingleAllocResultDurationName   = "controller.allocator.sync_single_alloc_result.duration"

	allocatorSuccessAttribute = "success"
)

// allocatorDurationBuckets covers API-call-scale allocation latencies, in seconds.
var allocatorDurationBuckets = []float64{0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5, 10}

var (
	allocatorScheduleDurationHistogram          metric.Float64Histogram
	allocatorPersistAllocStateDurationHistogram metric.Float64Histogram
	allocatorSyncAllocResultDurationHistogram   metric.Float64Histogram
	allocatorSyncSingleAllocResultHistogram     metric.Float64Histogram
)

func init() {
	meter := otel.GetMeterProvider().Meter(allocatorMeterName)
	allocatorScheduleDurationHistogram = newHistogram(meter, allocatorScheduleDurationMetricName,
		"Duration of allocator schedule decisions")
	allocatorPersistAllocStateDurationHistogram = newHistogram(meter, allocatorPersistAllocStateDurationMetricName,
		"Duration of persisting sandbox allocation state to the BatchSandbox annotation")
	allocatorSyncAllocResultDurationHistogram = newHistogram(meter, allocatorSyncAllocResultDurationMetricName,
		"Duration of syncing the batch allocation result to all sandboxes")
	allocatorSyncSingleAllocResultHistogram = newHistogram(meter, allocatorSyncSingleAllocResultDurationName,
		"Duration of syncing the allocation result for a single sandbox")
}

// newHistogram degrades to a no-op instrument on failure: telemetry must never
// take down controller startup.
func newHistogram(meter metric.Meter, name, description string) metric.Float64Histogram {
	histogram, err := meter.Float64Histogram(
		name,
		metric.WithDescription(description),
		metric.WithUnit("s"),
		metric.WithExplicitBucketBoundaries(allocatorDurationBuckets...),
	)
	if err != nil {
		logf.Log.Error(err, "telemetry: failed to create histogram, using no-op", "name", name)
		noopp, _ := noop.NewMeterProvider().Meter("noop").Float64Histogram(name)
		return noopp
	}
	return histogram
}

func recordAllocatorScheduleDuration(ctx context.Context, pool *sandboxv1alpha1.Pool, duration time.Duration, err error) {
	allocatorScheduleDurationHistogram.Record(ctx, duration.Seconds(), metric.WithAttributes(
		attribute.String("namespace", pool.Namespace),
		attribute.String("pool_name", pool.Name),
		attribute.Bool(allocatorSuccessAttribute, err == nil),
	))
}

func recordAllocatorPersistAllocStateDuration(ctx context.Context, sandbox *sandboxv1alpha1.BatchSandbox, duration time.Duration, err error) {
	allocatorPersistAllocStateDurationHistogram.Record(ctx, duration.Seconds(), metric.WithAttributes(
		attribute.String("namespace", sandbox.Namespace),
		attribute.String("pool_name", sandbox.Spec.PoolRef),
		attribute.Bool(allocatorSuccessAttribute, err == nil),
	))
}

func recordAllocatorSyncAllocResultDuration(ctx context.Context, pool *sandboxv1alpha1.Pool, duration time.Duration, err error) {
	allocatorSyncAllocResultDurationHistogram.Record(ctx, duration.Seconds(), metric.WithAttributes(
		attribute.String("namespace", pool.Namespace),
		attribute.String("pool_name", pool.Name),
		attribute.Bool(allocatorSuccessAttribute, err == nil),
	))
}

func recordAllocatorSyncSingleAllocResultDuration(ctx context.Context, sandbox *sandboxv1alpha1.BatchSandbox, duration time.Duration, err error) {
	allocatorSyncSingleAllocResultHistogram.Record(ctx, duration.Seconds(), metric.WithAttributes(
		attribute.String("namespace", sandbox.Namespace),
		attribute.String("pool_name", sandbox.Spec.PoolRef),
		attribute.Bool(allocatorSuccessAttribute, err == nil),
	))
}
